AtomMapper base classes

dwhswenson opened the pull request:
Resolves #5.

This implements two abstract classes for atom mappers:

* `AtomMapper`: This is a pretty trivial ABC.
* `ScoredAtomMapper`: This ABC combines `AtomMapper` with a scoring function. See discussion in [#6(comment)](https://github.com/OpenFreeEnergy/openfe/issues/6#issuecomment-1023500766) and related for the inspiration.

For both, implementers must implement `_mappings_generator`. For `ScoredAtomMapper`, implementers must also implement `_calculate_score`.

Note that this PR includes #29, so that one needs to be merged first. Once that's merged I'll pull those changes in here so that the diff doesn't include stuff from #29. (Or I think you can just look at 314430723a4cb3ab3eaccc994160b9df6a8ba1ce).

pep8speaks commented:
Hello @dwhswenson! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`openfe/setup/atommapper.py`](https://github.com/OpenFreeEnergy/openfe/blob/d6d89a914bf4e42690dd6c6328e4397d2fab34da/openfe/setup/atommapper.py):

> [Line 11:1](https://github.com/OpenFreeEnergy/openfe/blob/d6d89a914bf4e42690dd6c6328e4397d2fab34da/openfe/setup/atommapper.py#L11): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file







* In the file [`openfe/tests/setup/test_atommapper.py`](https://github.com/OpenFreeEnergy/openfe/blob/d6d89a914bf4e42690dd6c6328e4397d2fab34da/openfe/tests/setup/test_atommapper.py):

> [Line 8:1](https://github.com/OpenFreeEnergy/openfe/blob/d6d89a914bf4e42690dd6c6328e4397d2fab34da/openfe/tests/setup/test_atommapper.py#L8): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

richardjgowers reviewed (changes requested):
I'm not sure we need the `ScoredAtomMapper` abstract class, but it does serve as a useful example.  It's very possible that there's a mapper with 2 scoring functions, so usage alone the lines of:


```python
class MapperWithScorers:
   def suggest_mappings(self, mol1, mol2):
        ...

    def criteria_one(self, mapping):
        return something fancy

    def criteria_two(self, mapping):
        return something else

my_mapper = MapperWithScorers()

network = generate_network(mappers=[my_mapper], scorers=[my_mapper.criteria_one, my_mapper.criteria_two], weights=[0.4, 0.6])
```

dwhswenson commented:
@richardjgowers : Agreed, updated. I was mainly trying to keep the spirit of having the `score` method in the `AtomMapper` as issue #5 suggested. It didn't make sense to leave that as an optionally-implemented method as the existence of `Scorer`s implied. So I added the ideas from https://github.com/OpenFreeEnergy/openfe/issues/6#issuecomment-1023500766.

Tests for `ScoredAtomMapper` were a lot of fun, but hey -- sometimes ya gotta murder your darlings.

Current implementation still separates `_mappings_generator` as the internally-implemented thing from `suggest_mappings` as the user-facing method. This was done to keep consistency when working with subclasses of `AtomMapper`. Keep it this way or kill it?

Also, as class with nothing but a single public method, is there a need for `AtomMapper` to be class, rather than just any callable? I'm not sure which approach is more straightforward for inexperienced devs.

richardjgowers commented:
@dwhswenson yeah could just make this any callable too, will discuss.  Making people implement a private method makes the top level documentation consistent (as this is from the base class).
